nightfall-client/src/services/commitment-storage.mjs

```diff
@@ -609,147 +670,349 @@ async function findUsableCommitments(compressedZkpPublicKey, ercAddress, tokenId
+        change = tempChange;
+        commitmentsToUse = [commitmentsFiltered[i], ...threeCommitmentSum];
+      }
+    }
+  }
+
+  return commitmentsToUse;
+}
+
+/**
+ * Given an array of commitments, tries to find a subset of N elements
+ * whose sum is equal or higher than the target value
+ */
+function getSubset(commitments, value, N) {
+  let subset = [];
+  if (N === 1) {
+    subset = findSubsetOneCommitment(commitments, value);
+  } else if (N === 2) {
+    subset = findSubsetTwoCommitments(commitments, value);
+  } else if (N === 3) {
+    subset = findSubsetThreeCommitments(commitments, value);
+  } else if (N === 4) {
+    subset = findSubsetFourCommitments(commitments, value);
   }
-  return null;
+
+  return subset;
+}
+
+async function findUsableCommitments(
+  compressedZkpPublicKey,
+  ercAddress,
+  tokenId,
+  ercAddressFee,
+  _value,
+  _fee,
+) {
+  const value = generalise(_value); // sometimes this is sent as a BigInt.
+  const fee = generalise(_fee); // sometimes this is sent as a BigInt.
+
+  const commitmentsVerification = await verifyEnoughCommitments(
+    compressedZkpPublicKey,
+    ercAddress,
+    tokenId,
+    value,
+    ercAddressFee,
+    fee,
+  );
+
+  if (!commitmentsVerification) return null;
+
+  const { commitments, minC, minFc, commitmentsFee } = commitmentsVerification;
+
+  logger.debug(
+    `The user has ${commitments.length} commitments and needs to use at least ${minC} commitments to perform the transfer`,
+  );
+
+  if (fee.bigInt > 0n) {
+    logger.debug(
+      `The user has ${commitmentsFee.length} commitments and needs to use at least ${minFc} commitments to perform the transfer`,
+    );
+  }
+
+  const possibleSubsetsCommitments = [];
+
+  // Get the "best" subset of each possible size to then decide which one is better overall
+  // From the calculations performed in "verifyEnoughCommitments" we know that at least
+  // minC commitments are required. On the other hand, we can use a maximum of 4 commitments
+  // but we have to take into account that some spots needs to be used for the fee and that
+  // maybe the user does not have as much commitments
+  for (let i = minC; i <= min(commitments.length, 4 - minFc); ++i) {
+    const subset = getSubset(commitments, value, i);
+    possibleSubsetsCommitments.unshift(subset);
+  }
+
+  // Rank the possible commitments subsets.
+  // We prioritize the subset that minimizes the change.
+  // If two subsets have the same change, we priority the subset that uses more commitments
+  const rankedSubsetCommitmentsArray = possibleSubsetsCommitments
+    .filter(subset => subset.length > 0)
+    .sort((a, b) => {
+      const changeA = a.reduce((acc, com) => acc + com.preimage.value.bigInt, 0n) - value.bigInt;
+      const changeB = b.reduce((acc, com) => acc + com.preimage.value.bigInt, 0n) - value.bigInt;
+      if (changeA - changeB === 0n) {
+        return b.length - a.length;
+      }
+
+      return Number(changeA - changeB);
```

Comment:
I don't think this is safe for `BigInt` due to overflows, (although changes aren't likely to overflow).

This is probably safer `return changeA > change B ? 0 : -1`